058-kraft-upgrades-and-downgrades.md

```diff
@@ -0,0 +1,221 @@
+* Use the Kafka Admin API or CLI tools to do the unsafe downgrade
+* Unpause the reconciliation of the Kafka CR
+
+##### Configuring the initial metadata version
+
+When creating a new Kafka cluster, the initial metadata version can be configured using the `--release-version` option of the `kafka-storage.sh` script.
+To allow users to deploy Kafka clusters with older metadata version, Strimzi will also store the desired `metadata.version` in the per-broker configuration Config Map. 
+The value will be mounted into the Kafka pods and used when formatting the storage of a new Kafka cluster.
+That way - unlike when using environment variable - we will not need to roll the Pod every time the `metadata.version` changes (since in such cases we can update it dynamically without rolling the pods).
+
+#### Upgrade procedure
+
+Upgrade can happen in two situations:
+* After the upgrade of the Strimzi Cluster Operator:
+    * When the new version of the Cluster Operator supports a new default Kafka version and `Kafka.spec.kafka.version` is not set.
+    * When the user changes `Kafka.spec.kafka.version` in parallel with the Strimzi Cluster Operator upgrade.
+* When user requests Kafka upgrade by changing `Kafka.spec.kafka.version`.
+
+In both cases, the Strimzi Cluster Operator will:
+1. Validate that the metadata version currently used is compatible with the new Kafka version (it will check that it is lower than the new Kafka version).
+   This check is expected to pass for upgrades since a Kafka cluster that would not pass it would be invalid already before the upgrade.
+2. Roll all Kafka pods to use the containers with the new Kafka version.
+
+Once all the Kafka pods are rolled, the next step will depend on whether `Kafka.spec.kafka.metadataVersion` is set or not.
+When it is not set, Strimzi will automatically update the metadata version in the Kafka cluster using Kafka Admin API to the default version corresponding to the new Kafka version.
+The operator will not change the `Kafka.spec.kafka.metadataVersion` field - it will remain unset.
+And the Kafka upgrade will be complete with this.
+
+In case the `Kafka.spec.kafka.metadataVersion` field is set, it will just check the Kafka cluster has the desired metadata version as in any other reconciliation.
+The user will be expected to verify the Kafka cluster and change the `Kafka.spec.kafka.metadataVersion` when everything seems to work fine.
+Only after changing the metadata version, the upgrade will be considered complete.
+Similarly to today's ZooKeeper-based implementation, a warning will be issued (in log and in Kafka CR `.status` section) when the metadata version does not correspond to the Kafka version and tell the user to complete the upgrade by updating the metadata version.
+
+The upgrades will be supported to skip multiple Strimzi and Kafka versions in single step.
```

Comment:
It means that Kafka will support metadata upgrade to skip multiple versions as well?

Reply:
Yes, that is correct.